versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/codecs/BsonToEntityConverter.java

```diff
@@ -0,0 +1,158 @@
+          attributes.put(name, readArray(reader));
+          break;
+        case BOOLEAN:
+          attributes.put(name, Entity.ofBoolean(reader.readBoolean()));
+          break;
+        case STRING: {
+          final String value = reader.readString();
+          if (value.startsWith(MongoDbConstants.STRING_PREFIX)) {
+            attributes.put(name, Entity.ofString(value.substring(1)));
+          } else {
+            attributes.put(name, Entity.ofNumber(value.substring(1)));
+          }
+          break;
+        }
+        case BINARY:
+          attributes.put(name, Entity.ofBinary(reader.readBinaryData().getData()));
+          break;
+        default:
+          throw new UnsupportedOperationException(
+            String.format("Unsupported BSON type: %s", reader.getCurrentBsonType().name()));
+      }
+    }
+
+    reader.readEndDocument();
+    return attributes;
+  }
+
+  /**
+   * Read a BSON array and deserialize to associated Entity types.
+   * @param reader the reader to deserialize the Entities from.
+   * @return The deserialized list-type Entity, either a LIST or STRING_SET.
+   */
+  private static Entity readArray(BsonReader reader) {
+    reader.readStartArray();
+    if (reader.readBoolean()) {
+      // The entity is a generic list.
+      final List<Entity> entities = new ArrayList<>();
+
+      while (BsonType.END_OF_DOCUMENT != reader.readBsonType()) {
+        switch (reader.getCurrentBsonType()) {
+          case DOCUMENT:
+            entities.add(Entity.ofMap(readDocument(reader)));
+            break;
+          case ARRAY:
+            entities.add(readArray(reader));
+            break;
+          case BOOLEAN:
+            entities.add(Entity.ofBoolean(reader.readBoolean()));
+            break;
+          case STRING: {
+            final String value = reader.readString();
+            if (value.startsWith(MongoDbConstants.STRING_PREFIX)) {
+              entities.add(Entity.ofString(value.substring(1)));
+            } else {
+              entities.add(Entity.ofNumber(value.substring(1)));
+            }
+            break;
+          }
+          case BINARY:
+            entities.add(Entity.ofBinary(reader.readBinaryData().getData()));
+            break;
+          default:
+            throw new UnsupportedOperationException(
+              String.format("Unsupported BSON type: %s", reader.getCurrentBsonType().name()));
+        }
+      }
+
+      reader.readEndArray();
+      return Entity.ofList(entities);
+    }
+
+    // The entity is a string set, so simply read strings from it.
+    final Set<String> entities = new HashSet<>();
+    while (BsonType.END_OF_DOCUMENT != reader.readBsonType()) {
+      if (BsonType.STRING != reader.getCurrentBsonType()) {
+        throw new UnsupportedOperationException(
+          String.format("STRING_SET not composed of STRING values, found type: %s", reader.getCurrentBsonType().name()));
+      }
+
+      entities.add(reader.readString());
```

Comment:
Per comments elsewhere, this is a place that reinforces the hackiness of the string prefixing. What happens if we introduce a number set?